### PR TITLE
Contact jwt

### PR DESCRIPTION
**Pull request by alari**

### CODE REVIEW

File: kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+          } yield (header, claim))
+
+          pk ← EitherT.fromEither(getPk(hc._1, hc._2))
+
+          _ ← checker.check(Signature(pk, ByteVector(s.getBytes)), ByteVector((h + c).getBytes())).leftMap(err ⇒ err: Throwable) // TODO: coproduct
+
+        } yield hc
+
+      case _ ⇒
+        EitherT.leftT(new IllegalArgumentException("Token malformed: must contain exactly two dots")) // TODO: coproduct
```
Comment (RicoGit) [inline kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala]: Nice to have 'invalid token' in a error message for seeing what is actually problem is.

File: kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+   */
+  def read[F[_] : Monad, H : Decoder, C : Decoder](
+    token: String,
+    getPk: (H, C) ⇒ Either[NoSuchElementException, KeyPair.Public],
+    checker: SignatureChecker): EitherT[F, Throwable, (H, C)] =
+    token.split('.').toList match {
+      case h :: c :: s :: Nil ⇒
+
+        for {
+          hbv ← EitherT.fromOption(ByteVector.fromBase64(h, alphabet), new IllegalArgumentException("Can't read base64 header"))
```
Comment (RicoGit) [inline kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala]: Nice to have 'invalid header' in a error message for seeing what is actually problem is.

File: kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          cbv ← EitherT.fromOption(ByteVector.fromBase64(c, alphabet), new IllegalArgumentException("Can't read base64 claim"))
+
+          hc ← EitherT.fromEither[F](for {
+            hj ← parse(new String(hbv.toArray))
+            cj ← parse(new String(cbv.toArray))
+            header ← Decoder[H].decodeJson(hj)
+            claim ← Decoder[C].decodeJson(cj)
```
Comment (RicoGit) [inline kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala]: Is methods 'parse' and 'decodeJson' can't throw any exception?

Comment (alari) [reply kademlia/protocol/src/main/scala/fluence/kad/protocol/Jwt.scala]: Yep, it returns Either

File: kademlia/protocol/src/main/scala/fluence/kad/protocol/Contact.scala [hunk 4]
```diff
@@ -54,42 +81,58 @@ case class Contact(
...
-    )
+  case class JwtData(ip: InetAddress, port: Int, gitHash: String)
 
-    val inetAddress = bytes(4).xmap[InetAddress](
-      bv ⇒ InetAddress.getByAddress(bv.toArray),
-      ia ⇒ ByteVector(ia.getAddress)
-    )
+  object JwtImplicits {
+    implicit val encodeHeader: Encoder[JwtHeader] = header ⇒ Json.obj(
+      "pk" -> Json.fromString(header.publicKey.value.toBase64),
```
Comment (DieMyst) [inline kademlia/protocol/src/main/scala/fluence/kad/protocol/Contact.scala]: base64 -> url alphabet?

File: kademlia/protocol/src/main/scala/fluence/kad/protocol/Contact.scala [hunk 5]
```diff
@@ -54,42 +81,58 @@ case class Contact(
...
-      {
-        case (((v, ia), p), gh) ⇒
-          Contact(ip = ia, port = p, protocolVersion = v, gitHash = gh)
-      },
-      c ⇒ (((c.protocolVersion, c.ip), c.port), c.gitHash)
+    implicit val decodeHeader: Decoder[JwtHeader] = c ⇒
+      for {
+        pk ← c.downField("pk").as[String]
+        pv ← c.downField("pv").as[Long]
+        pubKey ← ByteVector.fromBase64(pk).fold[Either[DecodingFailure, KeyPair.Public]](
```
Comment (DieMyst) [inline kademlia/protocol/src/main/scala/fluence/kad/protocol/Contact.scala]: base64 -> url alphabet?

File: kademlia/protocol/src/main/scala/fluence/kad/protocol/Contact.scala (hunk 4 above)
Comment (alari) [reply kademlia/protocol/src/main/scala/fluence/kad/protocol/Contact.scala]: I'm not sure if it matches the JWT spec. I'll take a short look

**Pull request merged**